pkg/spegel/spegel.go

```diff
@@ -0,0 +1,261 @@
+package spegel
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/k3s-io/k3s/pkg/clientaccess"
+	"github.com/k3s-io/k3s/pkg/daemons/config"
+	"github.com/k3s-io/k3s/pkg/version"
+	"github.com/rancher/dynamiclistener/cert"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/request/union"
+	"k8s.io/utils/pointer"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/stdr"
+	"github.com/gorilla/mux"
+	leveldb "github.com/ipfs/go-ds-leveldb"
+	ipfslog "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/p2p/host/peerstore/pstoreds"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/xenitab/spegel/pkg/metrics"
+	"github.com/xenitab/spegel/pkg/oci"
+	"github.com/xenitab/spegel/pkg/registry"
+	"github.com/xenitab/spegel/pkg/routing"
+	"github.com/xenitab/spegel/pkg/state"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+// DefaultRegistry is the default instance of a Spegel distributed registry
+var DefaultRegistry = &Config{
+	Bootstrapper: NewSelfBootstrapper(),
+	HandlerFunc: func(_ *Config, _ *mux.Router) error {
+		return errors.New("not implemented")
+	},
+	AuthFunc: func() authenticator.Request {
+		return union.New(nil)
+	},
+}
+
+var (
+	P2pAddressAnnotation = "p2p." + version.Program + ".cattle.io/node-address"
+	P2pEnabledLabel      = "p2p." + version.Program + ".cattle.io/enabled"
```

Comment:
Do you foresee a case where the node has the p2p node-address annotation filled but the enabled label is false? Just wondering if we need the annotation and the label

Reply:
The label is used for matching when listing nodes using a label selector. You can't match/filter nodes based on annotations, and I wanted to be able to select only nodes that have this enabled.

When I started tinkering with this, the embedded registry was a node-level option instead of cluster-level, and I left the node selector logic here in case we do in the future want to allow some nodes to opt out of using the embedded registry.